agent/containermetadata/metadata_manager.go

```diff
@@ -0,0 +1,131 @@
+// operations
+type MetadataManager interface {
+	SetContainerInstanceArn(string)
+	CreateMetadata([]string, *api.Task, *api.Container) ([]string, error)
+	UpdateMetadata(string, *api.Task, *api.Container) error
+	CleanTaskMetadata(*api.Task) error
+}
+
+// metadataManager implements the MetadataManager interface
+type metadataManager struct {
+	client               dockerMetadataClient
+	cfg                  *config.Config
+	containerInstanceARN string
+}
+
+// NewMetadataManager creates a metadataManager for a given DockerTaskEngine settings.
+func NewMetadataManager(client dockerMetadataClient, cfg *config.Config) MetadataManager {
+	return &metadataManager{
+		client: client,
+		cfg:    cfg,
+	}
+}
+
+// SetContainerInstanceArn sets the metadataManager's ContainerInstanceArn which is not available
+// at its creation as this information is not present immediately at the agent's startup
+func (manager *metadataManager) SetContainerInstanceArn(containerInstanceARN string) {
+	// Do nothing if disabled
+	if !manager.cfg.ContainerMetadataEnabled {
+		return
+	}
+	manager.containerInstanceARN = containerInstanceARN
+}
+
+// UpdateMetadata updates the metadata file after container starts and dynamic
+// metadata is available
+func (manager *metadataManager) UpdateMetadata(dockerID string, task *api.Task, container *api.Container) error {
+	// Do nothing if disabled
+	if !manager.cfg.ContainerMetadataEnabled {
+		return nil
+	}
+
+	// Do not update (non-existent) metadata file for internal containers
+	if container.IsInternal {
+		return nil
+	}
+
+	// Verify metadata file exists before proceeding
+	if !metadataFileExists(task, container, manager.cfg.DataDir) {
+		expectedPath, _ := getMetadataFilePath(task, container, manager.cfg.DataDir)
+		return fmt.Errorf("container metadata update: %s does not exist", expectedPath)
+	}
+
+	// Get docker container information through api call
+	dockerContainer, err := manager.client.InspectContainer(dockerID, inspectContainerTimeout)
+	if err != nil {
+		return err
+	}
+
+	// Ensure we do not update a container that is invalid or is not running
+	if dockerContainer == nil || !dockerContainer.State.Running {
+		return fmt.Errorf("container metadata update: container not running or invalid")
+	}
+
+	// Get last metadata file creation time
+	createTime, err := getMetadataFileUpdateTime(task, container, manager.cfg.DataDir)
+	if err != nil {
+		seelog.Errorf("container metadata update: %v", err)
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	var updateTime time.Time
+	metadata := manager.parseMetadata(createTime, updateTime, dockerContainer, task)
+	err = metadata.writeToMetadataFile(task, container, manager.cfg.DataDir)
+	if err != nil {
+		return err
+	}
+
+	// Update the file again with time stamp of last file update. This will cause the file to have
+	// a different time stamp from its actual file stat inspection but this is unavoidable in our
```

Comment:
I don't think we care about the stat time.  Let's just write the file once instead of several times.